lib/index.js

```diff
@@ -12,34 +26,312 @@ class PluginPassportOAuth {
+          constructor: this.strategyConstructor,
+          strategyOptions: this.config.strategies[strategy].credentials,
+          authenticateOptions: {
+            scope: this.config.strategies[strategy].scope
+          },
+          fields: ['access_token', 'refresh_token']
+        },
+        methods: {
+          create: 'create',
+          delete: 'delete',
+          exists: 'exists',
+          getInfo: 'getInfo',
+          update: 'update',
+          validate: 'validate',
+          verify: 'verify',
+          afterRegister: 'afterRegister'
+        }
+      };
+    });
+  }
+
+  /**
+   * Get the repository according to the strategy
+   *
+   * @param provider
+   * @returns {module.exports.constructors.Repository|*}
+   */
+  getProviderRepository (provider) {
+    if (!this.providerRepository) {
+      this.providerRepository = new this.context.constructors.Repository(provider);
+    }
+
+    return this.providerRepository;
+  }
+
+  /**
+   * Get the credentials of a given kuid from the repository of the corresponding strategy
+   *
+   * @param kuid
+   * @param provider
+   * @returns {Promise|Promise.<TResult>}
+   */
+  getCredentialsFromKuid (kuid, provider) {
+    return this.getProviderRepository(provider).search({
+      query: {
+        match: {
+          kuid
+        }
+      }
+    })
+      .then(result => {
+        if (result.total === 0) {
+          return Promise.resolve(null);
+        }
+
+        return Promise.resolve(result.hits[0]);
+      });
+  }
+
+  /**
+   * Verify function called by passport to log the user.
+   * If a persist option is in the configuration of the strategy then every field present in this configuration will be persisted in the repository.
+   *
+   * @param request
+   * @param accessToken
+   * @param refreshToken
+   * @param profile
+   * @returns {Promise.<T>}
+   */
+  verify (request, accessToken, refreshToken, profile) {
+    const
+      user = {};
+
+    return this.getProviderRepository(profile.provider)
+      .get(profile._json[this.config.strategies[profile.provider].identifierAttribute])
+      .then(userObject => {
+        let req = null;
+
+        if (userObject !== null) {
+          return Promise.resolve(userObject.kuid);
+        }
+
+        if (this.config.strategies[profile.provider].persist.length !== 0) {
+          Object.keys(profile._json).forEach(attr => {
+            if (this.config.strategies[profile.provider].persist.indexOf(attr) > -1) {
+              user[attr] = profile._json[attr];
+            }
+          });
+
+          req = {
+            controller: 'security',
+            action: 'createUser',
+            body: {
+              content: {
+                profileIds: this.config.defaultProfiles ? this.config.defaultProfiles : ['default']
+              }, // content will be persisted in Kuzzle
+              credentials: {
+                kuid: profile._json[this.config.strategies[profile.provider].identifierAttribute]
```

Comment:
not needed, `kuid` will be set when Kuzzle will call plugin's `create` method (with the good Kuzzle uid)